guardian/mobile-n10n
# Clean unregistered android tokens

## What does this change?

When attempting to send android users a notification, this changes which error responses from firebase are parsed as an `InvalidToken`. Parsing error responses as an `InvalidToken` will trigger android tokens to be cleaned from the registrations db.

We can see in the logs that we have a large number of errors from the android sender lambda (where peaks in these error messages occur when sending breaking news notifications to a large number of recipients):
![Screenshot 2022-09-01 at 12 40 04](https://user-images.githubusercontent.com/45561419/187905353-ba658b51-c93c-471a-953c-1a80ca844b33.png)

We can see in the google console that these error responses have been present for over a month:
![Screenshot 2022-09-01 at 12 40 42](https://user-images.githubusercontent.com/45561419/187905452-507f6171-d361-4cbc-aa2a-b9429c9c0142.png)

We can see that ios sender lambdas trigger about 1.7k tokens a week to be cleaned:
![Screenshot 2022-09-01 at 12 41 45](https://user-images.githubusercontent.com/45561419/187905631-406f65d4-2048-4d72-bc27-06538f25ddda.png)


Whereas android sender lambdas trigger only about 10 tokens a week to be cleaned (significantly less, even accounting for the difference in android vs ios users):
![Screenshot 2022-09-01 at 12 41 19](https://user-images.githubusercontent.com/45561419/187905567-d52a4596-92db-48ab-95df-cccf7e46c2b2.png)

Looking at the errors in the android sender lambdas in more detail, we can see that the error responses all share some common data:
- error code: `NOT_FOUND`
- error cause: `Requested entity was not found.`
- stack trace includes message:

```
Caused by: com.google.api.client.http.HttpResponseException: 404
POST https://fcm.googleapis.com/v1/projects/push-notifications-prod-15cbe/messages:send
{
  "error": {
    "code": 404,
    "message": "Requested entity was not found.",
    "status": "NOT_FOUND",
    "details": [
      {
        "@type": "type.googleapis.com/google.firebase.fcm.v1.FcmError",
        "errorCode": "UNREGISTERED"
      }
    ]
  }
}
```

## How to test

- We have provided unit tests to cover some of the pre-existing functionality as well as the new logic implemented.

In CODE we can test by:
- inserting a registration into the test rig db for a topic with no other subscribers
- trigger a notification send to the given topic in code, without using dryRun
- confirm that the token is successfully retrieved from the db and is attempted to be sent by the android sender lambda
- inspect the response from firebase and confirm, where the response information matches the log messages we've seen in prod, that the token is successfully parsed as an `InvalidToken` and that it is eventually cleaned from the db. 

## How can we measure success?

- Successful responses from firebase continue to be parsed as expected
- More tokens are parsed as invalid and successfully cleaned from the db
- Number of errors in our logs decreases
- Number of error responses shown in the google console also decrease

## Have we considered potential risks?

We've tried to mitigate the risk by introducing unit tests covering both pre-existing and new functionality. We will also test in code to confirm the behaviour is expected.

We will also attempt to confirm our understanding of the error code with google/firebase before pushing this change (TBC).
